File: subsets/subsets.cpp
class Solution {
public:

   void solve(vector<int> nums,vector<int>output,int index,vector<vector<int>> &ans){

     // Base case

     if(index>=nums.size()){
         ans.push_back(output);
         return;
     }

     // Exclude

     solve(nums,output,index+1,ans);

     // Include

     output.push_back(nums[index]);
     solve(nums,output,index+1,ans);



   }
   

    vector<vector<int>> subsets(vector<int>& nums) {

        vector<vector<int>>ans;
        vector<int>output;

        solve(nums,output,0,ans);
        return ans;
        
    }
};

File: 78-subsets/subsets.java
public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> subsets = new ArrayList<>();
        generateSubsets(nums, 0, new ArrayList<Integer>(), subsets);
        return subsets;
    }

private void generateSubsets(int[] nums, int index, List<Integer> current, List<List<Integer>> subsets) {
        if (index == nums.length) {
            subsets.add(new ArrayList<>(current));
            return;
        }

        // Include the current number
        current.add(nums[index]);
        generateSubsets(nums, index + 1, current, subsets);
        current.remove(current.size() - 1);

        // Exclude the current number
        generateSubsets(nums, index + 1, current, subsets);
    }
}


